### ISSUE TITLE
Various issues with ridge filters.

### ISSUE THREAD

**Issue by anntzer** [scikit-image/scikit-image#6436]
# Description

Unless I am mistaken, *all four* ridge filters implemented in scikit-image (https://scikit-image.org/docs/dev/auto_examples/edges/plot_ridge_filter.html) differ from the algorithms described in the corresponding papers (each linked in the functions' docstrings).
In the writeup below, I tried to collect the relevant equations from the original papers, and list the issues that exist in scikit-image's implementations.  Some of them are relatively minor, but others (marked with a (!)) are significant deviations from the original filters.

I apologize in advance for any possible misreadings of the articles or implementations, and welcome proofreading of my writeup.

If we agree that these deviations are indeed incorrect, I can perhaps try to provide fixes for them, although back-compatibility would need to be discussed. 

----------

We note

-   $\lambda_i$ the Hessian eigenvalues (see also #5561/#6149 for the computation of the Hessian, which has been fixed for `meijering()` but not the other filters);
-   $f(\mathbf{x})$ the final filtered image;
-   ${(x)}^+ = \max(x, 0)$ (the positive part of $x$).

Most papers only define the filters for bright ridges ($\lambda < 0$). We rearrange the formulas to use $(-\lambda)$ where possible with the idea that these terms are positive. We also derive the corresponding formulas for dark ridges ($\lambda > 0$) by changing the signs of the eigenvalues and swapping minima and maxima as needed.

## 2D Meijering filter

[Meijering *et al.*, *Cytometry Part A* (2004).](https://doi.org/10.1002/cyto.a.20022) [Implementation](https://github.com/ImageScience/NeuronJ/blob/master/source/java/NeuronJ_.java).

The Hessian is computed by convolution with second Gaussian derivatives. (This is not entirely clear in the text, but one can check the implementation of `imagescience.feature.Differentiator`.)

Let $\lambda_i' = \lambda_i + \alpha\sum_{j\not=i}\lambda_j$ (where $\alpha$ defaults to $-1/3$), sorted by increasing magnitude (absolute value). (Which of $\lambda_i$ or $\lambda_i'$ get sorted by magnitude is unclear in the article, but the implementation directly computes $\lambda_i'$ (which it calls `L1` and `L2`) for $\alpha=-1/3$ and sorts their magnitudes.) $\lambda_d'$ is the greatest one. Then, for bright ridges ($\lambda < 0$), $$f(\mathbf{x}) = {\left(\frac{-\lambda_d'(\mathbf{x})}{\max_{\mathbf{x}}(-\lambda_d'(\mathbf{x}))}\right)}^+.$$

Equivalently, for dark ridges ($\lambda > 0$), $$f(\mathbf{x}) = {\left(\frac{\lambda_d'(\mathbf{x})}{\max_{\mathbf{x}}\lambda_d'(\mathbf{x})}\right)}^+.$$

A single scale $\sigma$ is used.

### scikit-image status:

The implementation is based on the bright ridges case ($\lambda < 0$).

-  (!) Sorting is done by $\left|\lambda_i\right|$, not $\left|\lambda'_i\right|$ (the order may not be the same).
- The filter is only designed for the 2D case (with a specific further post-processing in mind); generalization to 3D is an addition by scikit-image (this is OK but perhaps warrants a note).  @efmkoene notes in https://github.com/scikit-image/scikit-image/issues/5561#issuecomment-920997906 that the optimal value of $\alpha$ is still $-1/3$ for the 3D case, although I haven't checked the derivation in that case.
-   The filter is designed for a single $\sigma$ value and the article does not discuss whether maximizing over values obtained for various $\sigma$ makes sense, or whether they should be normalized before comparison.

## 3D Sato filter

[Sato *et al.*, *Medical Image Analysis* (1998).](https://doi.org/10.1016/s1361-8415(98)80009-1)

The Hessian is explicitly computed by convolution with second Gaussian derivatives (equation (5)).

For bright ridges ($\lambda < 0$), equation (9) gives $$f(\mathbf{x}) = ({(-\lambda_2)}^+{(-\lambda_3)}^+)^{1/2}$$ where $\lambda_2$ and $\lambda_3$ are the minimal (typically the most negative) eigenvalues.

Equivalently, for dark ridges ($\lambda > 0$), $$f(\mathbf{x}) = (\lambda_2^+\lambda_3^+)^{1/2}$$ where $\lambda_2$ and $\lambda_3$ are the maximal (typically the most positive) eigenvalues.

Maximization of $f$ over $\sigma$ is explicitly discussed.

#### scikit-image status:

The implementation is based on the dark ridges case ($\lambda > 0$).

-   (!) scikit-image computes ${\left|\lambda_2\lambda_3\right|}^{1/2}$, fixing the value to zero if $\lambda_3<0$, but not if $\lambda_2<0$.
-   The filter is only designed for the 3D case; generalization to 2D is an addition by scikit-image (this is OK but perhaps warrants a note).
-   Prior to maximization over $\sigma$, $f$ must be normalized by multiplication by $\sigma^2$ (equation (22)), which happens to match the implementation of `compute_hessian_eigenvalues` with `use_gaussian_derivatives=False` (because $f$ is homogeneous to an eigenvalue).
-   The Hessian is computed by finite differences.

## 2D/3D Frangi filter

[Frangi *et al.*, *MICCAI* (1998).](https://doi.org/10.1007/BFb0056195)

The Hessian is explicitly computed by convolution with second Gaussian derivatives (equation (2)).

The eigenvalues are sorted by increasing magnitude.

For bright ridges ($\lambda < 0$), define the plateness $A$, the blobness $B$, and the structuredness $S$ as $$A_{3D} = \frac{-\lambda_2}{-\lambda_3}; \qquad B_{3D} = \frac{\left|\lambda_1\right|}{\sqrt{{(-\lambda_2)}^+{(-\lambda_3)}^+}}; \qquad S_{3D} = \sqrt{\lambda_1^2+\lambda_2^2+\lambda_3^2}$$ $$A_{2D} = 0; \qquad B_{2D} = \frac{\left|\lambda_1\right|}{{(-\lambda_2)}^+}; \qquad S_{2D} = \sqrt{\lambda_1^2+\lambda_2^2}.$$

Equivalently, for dark ridges ($\lambda > 0$), $$A_{3D} = \frac{\lambda_2}{\lambda_3}; \qquad B_{3D} = \frac{\left|\lambda_1\right|}{\sqrt{\lambda_2^+\lambda_3^+}}; \qquad S_{3D} = \sqrt{\lambda_1^2+\lambda_2^2+\lambda_3^2}$$ $$A_{2D} = 0; \qquad B_{2D} = \frac{\left|\lambda_1\right|}{\lambda_2^+}; \qquad S_{2D} = \sqrt{\lambda_1^2+\lambda_2^2}.$$

Then $$f(\mathbf{x}) = \left[1-\exp\left(-\frac{A^2}{2\alpha^2}\right)\right]\exp\left(-\frac{B^2}{2\beta^2}\right)\left[1-\exp\left(-\frac{S^2}{2c^2}\right)\right]$$ where $\alpha = \beta = 1/2$ (default) and $c = \frac12\max_{\mathbf{x}}S_{\mathbf{x}}$ (default)[^1]. Note that the first factor always $1$ in the 2D case, and the second factor is zero when $\lambda_2$ (and $\lambda_3$) have the wrong sign.

Maximization of $f$ over $\sigma$ is explicitly discussed.

[^1]: This default $\sigma$-dependent $c$ is also used e.g. in [Oruganti *et al.* (2013)](https://doi.org/10.1117/12.2005988) and [Longo *et al.* (2020)](https://doi.org/10.1016/j.pacs.2020.100200).

#### scikit-image status:

The implementation is based on the bright ridges case ($\lambda < 0$).

-   (!!) The denominator of $B$ uses an absolute value instead of a positive part. The intent of the Frangi filter is that whenever an eigenvalue has the wrong sign, $f = 0$ (equation (13)), which is the same as using positive parts, because a wrong-sign $\lambda$ leads to $B = +\infty$ and $\exp(-B^2/2\beta^2) = 0$. Using absolute values completely loses this feature.
-   (!) The default value of $c$ is fixed to $15$, which is arbitrary, whereas the article explicitly calls out that this parameter must be optimized and suggests to initialize it to $\frac12\max_{\mathbf{x}}S_{\mathbf{x}}$. This is quite important, as the last factor is quite sensitive to scale mismatches.
-   The fact that $\alpha$ has no effect for 2D images could be mentioned in the docstring.
-   The Hessian is computed by finite differences.

## 2D hybrid Hessian filter

[Ng *et al.*, *ACCV* (2014).](https://doi.org/10.1007/978-3-319-16811-1_40)

This filter exactly matches the Frangi filter, but then specifically and only picks regions where $\max_\sigma f(\mathbf{x}) = 0$ (equation (15)).  However, this doesn't really make sense and doesn't correspond to figure 2(e) of the article, which seems to pick the complementary set ($\max_\sigma f(\mathbf{x}) > 0$).

#### scikit-image status:

- (!) scikit-image mixes up(?!) the Frangi filter values ($\mathcal{L}$ in the article) and the wrinkle mask values ($\mathcal{B}$ in the article, which corresponds exactly to the regions where $\max_\sigma f(\mathbf{x}) = 0$).

Ng et al. further postprocess the result of the binarization by only keeping regions greater than 250 pixels, but this seems to be something that scikit-image should anyways keep separate (as `morphology.remove_small_objects`).  But if my understanding of Ng et al. is correct, this doesn't warrant to be a separate function at all(?!).

# Version information
```python
# Paste the output of the following python commands
from __future__ import print_function
import sys; print(sys.version)
import platform; print(platform.platform())
import skimage; print(f'scikit-image version: {skimage.__version__}')
import numpy; print(f'numpy version: {numpy.__version__}')
```
```
3.10.5 (main, Jun  6 2022, 18:49:26) [GCC 12.1.0]
Linux-5.18.10-arch1-1-x86_64-with-glibc2.35
scikit-image version: 0.19.1
numpy version: 1.23.0
```

**Comment by efmkoene** [comment]
Thanks for opening the issue. [Prompted by your comment, I re-derived the alpha factor in the Meijering method](https://efmkoene.github.io/2022-07-14-Meijering/), and found that it actually generalizes to `alpha=-1/(1+dim)`. Thus, `alpha=-1/3` in 2D, `alpha=-1/4` in 3D, etc. So my original assertion that it was always -1/3 regardless of the dimension, wasn't correct!

Sorting by $\lambda$ instead of $\lambda'=\lambda_i+alpha\sum_{j\neq i}\lambda_j$ should be fine in practice, though; they follow the exact same order. You can rewrite the relation as $\lambda_i'=(1-\alpha)\lambda_i +\alpha \sum_j \lambda_j$. Then all $\lamba_i'$ have the same constant $\alpha \sum_j \lambda_j$ offset (thus not affecting the sorting), and we're just left with effective differences of $(1-\alpha)\lambda_i$ between the eigenvalues. As long as $\alpha <1$, which $\alpha$ should in principle follow (it should be negative!), the sorting remains the same.

**Comment by jni** [comment]
Thanks both for looking into these issues! I don't have time to check the math but it's very plausible indeed that mistakes crept into the original implementations here.

@anntzer regarding backward compatibility, in these cases scikit-image usually takes the policy of considering updated numerical results as bug fixes. We note them as API changes but we don't add deprecation warnings. So PRs with fixes would be most welcome!

Thank you very much for the thorough write up!

**Comment by anntzer** [comment]
Mostly handled by #6446, although the hessian() filter still seems wrong.

**Comment by grlee77** [comment]
I am going to close this issue as there is already a preceding issue related to the remaining issue with the `hessian` filter: #3783

**Comment by swinkelm** [comment]
It has been a while since this issue has been closed, but I came across these differences betweeen papers and implementations aswell and would like to discuss the implementation of the Sato filter.
As mentioned by @anntzer, equation (9) was implemented, but as far as I understand, this equation is only a base from which the more generalised equation (16) is derived. The latter takes three additional parameters which allow for much better results if set correctly.
I can be wrong about this of course, so I'd appreciate if someone could provide additional insights.
